### PR TITLE
Shopping cart

### PR DESCRIPTION
**Pull request by NikitaLafinskiy**

### CODE REVIEW

File: src/main/java/com/bookstore/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    @PreAuthorize("hasRole('USER')")
+    public ShoppingCartDto addCartItem(@AuthenticationPrincipal String email,
+                                       @RequestBody @Valid
+                                       AddBookCartRequestDto addBookCartRequestDto) {
+        return shoppingCartService.addCartItem(email, addBookCartRequestDto);
+    }
+
+    @PutMapping("/items/{cartItemId}")
+    @PreAuthorize("hasRole('USER')")
+    public CartItemDto updateCartItem(@RequestBody @Valid
```
Comment (sarakhmen) [inline src/main/java/com/bookstore/controller/ShoppingCartController.java]: do not move param on the next line here, fix style issues

Comment (sarakhmen) [reply src/main/java/com/bookstore/controller/ShoppingCartController.java]: fix the same issue in all places

File: src/main/java/com/bookstore/controller/ShoppingCartController.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+                                          UpdateCartRequestDto updateCartRequestDto,
+                                      @PathVariable Long cartItemId) {
+        return shoppingCartService.updateCartItem(updateCartRequestDto,
+                cartItemId);
+    }
+
+    @DeleteMapping("/items/{cartItemId}")
+    @PreAuthorize("hasRole('USER')")
```
Comment (sarakhmen) [inline src/main/java/com/bookstore/controller/ShoppingCartController.java:54]: return http status 204 here

File: src/main/java/com/bookstore/exception/CustomGlobalExceptionHandler.java [hunk 3]
```diff
@@ -42,6 +42,7 @@ public ResponseEntity<Object> handleRegistrationException(RegistrationException
     @ExceptionHandler(Exception.class)
     public ResponseEntity<Object> handleException(Exception ex) {
         Map<String, Object> error = new HashMap<>();
+        error.put("exceptionType", ex.getClass().getName());
```
Comment (sarakhmen) [inline src/main/java/com/bookstore/exception/CustomGlobalExceptionHandler.java:45]: why did you add changes not related to this HW to this PR? Let's do it in a separate PR and do not send them on review, if it's related to the bugfix/refactor etc

File: src/main/java/com/bookstore/repository/BookRepository.java [hunk 4]
```diff
@@ -10,6 +10,9 @@
 
 @Repository
 public interface BookRepository extends JpaRepository<Book, Long> {
-    @Query("select b from Book b inner join fetch b.categories c where c.id = :categoryId")
-    List<Book> findAllByCategoryId(Pageable pageable, @Param("categoryId") Long categoryId);
+    @Query("select b from Book b left join fetch b.categories c where c.id = :categoryId")
+    Page<Book> findAllByCategoryId(Pageable pageable, @Param("categoryId") Long categoryId);
+
+    @Query("select b from Book b left join fetch b.categories")
```
Comment (sarakhmen) [inline src/main/java/com/bookstore/repository/BookRepository.java]: let's use entitygraph here instead. Also, if you decide to write custom queries, let's make them more readable, use UPPERCASED keywords etc

File: src/main/java/com/bookstore/repository/ShoppingCartRepository.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+import com.bookstore.entity.ShoppingCart;
+import org.springframework.data.jpa.repository.EntityGraph;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
+    @EntityGraph(attributePaths = {"cartItems", "user", "cartItems.book"})
+    @Query("select c from ShoppingCart c where c.user.email = :email")
```
Comment (sarakhmen) [inline src/main/java/com/bookstore/repository/ShoppingCartRepository.java]: why did you use custom query if it's enough to follow JPA method name conventions and JPA will create the method with the supposed query itself? fix the same issue in all places

**Review by sarakhmen — CHANGES_REQUESTED** [review]
let's improve your solution :)

File: src/main/java/com/bookstore/controller/ShoppingCartController.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    public ShoppingCartDto getCart(@AuthenticationPrincipal String email) {
+        return shoppingCartService.getCart(email);
+    }
+
+    @PostMapping
+    @PreAuthorize("hasRole('USER')")
+    public ShoppingCartDto addCartItem(@AuthenticationPrincipal String email,
+                                       @RequestBody @Valid AddBookCartRequestDto
+                                               addBookCartRequestDto) {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/bookstore/controller/ShoppingCartController.java]: I can recommend you this parameter code style: 
```suggestion
    @PostMapping
    @PreAuthorize("hasRole('USER')")
    public ShoppingCartDto addCartItem(
        @AuthenticationPrincipal String email,
        @RequestBody @Valid AddBookCartRequestDto addBookCartRequestDto
    ) {
```

File: src/main/java/com/bookstore/entity/ShoppingCart.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @OneToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "user_id", nullable = false)
+    @MapsId
+    private User user;
+
+    @OneToMany(mappedBy = "shoppingCart",
+            fetch = FetchType.LAZY,
```
Comment (liliia-ponomarenko) [inline src/main/java/com/bookstore/entity/ShoppingCart.java]: It's the default fetch type for this relation

File: src/main/java/com/bookstore/mapper/ShoppingCartMapper.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    CartItemDto cartItemToDto(CartItem cartItem);
+
+    @Mapping(target = "book", source = "bookId", qualifiedByName = "bookById")
+    CartItem dtoToCartItem(CartItemDto cartItemDto);
+
+    @Mapping(target = "book", source = "bookId", qualifiedByName = "bookById")
+    CartItem dtoToCartItem(AddBookCartRequestDto addBookCartRequestDto);
+
+    void updateCartItem(UpdateCartRequestDto updateCartRequestDto,
+                        @MappingTarget CartItem cartItem);
```
Comment (liliia-ponomarenko) [inline src/main/java/com/bookstore/mapper/ShoppingCartMapper.java]: Better have a separate CartItemMapper

File: src/main/java/com/bookstore/service/cart/ShoppingCartServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        shoppingCart.addCartItem(cartItem);
+        shoppingCartRepository.save(shoppingCart);
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public CartItemDto updateCartItem(UpdateCartRequestDto updateCartRequestDto,
+                                      Long cartItemId) {
+        CartItem cartItem = cartItemRepository.findById(cartItemId)
+                .orElseThrow(EntityNotFoundException::new);
```
Comment (liliia-ponomarenko) [inline src/main/java/com/bookstore/service/cart/ShoppingCartServiceImpl.java]: Let's add an informative message

File: src/main/java/com/bookstore/exception/CustomGlobalExceptionHandler.java (hunk 3 above)
Comment (liliia-ponomarenko) [reply src/main/java/com/bookstore/exception/CustomGlobalExceptionHandler.java:45]: not fixed

File: src/main/resources/db/changelog/changes/05-create-cart-table.yaml [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+                  constraints:
+                    nullable: false
+        - addForeignKeyConstraint:
+            baseTableName: shopping_carts
+            baseColumnNames: user_id
+            constraintName: fk_shopping_carts_user
+            referencedTableName: users
+            referencedColumnNames: id
+
+  - changeSet:
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/05-create-cart-table.yaml]: better move it to separate file

File: src/main/resources/db/changelog/changes/05-create-cart-table.yaml [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+            baseColumnNames: book_id
+            constraintName: fk_cart_items_book
+            referencedTableName: books
+            referencedColumnNames: id
+        - addForeignKeyConstraint:
+            baseTableName: cart_items
+            baseColumnNames: shopping_cart_id
+            constraintName: fk_cart_items_shopping_cart
+            referencedTableName: shopping_carts
+            referencedColumnNames: user_id
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/05-create-cart-table.yaml]: leave one empty line at the end of the file

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution;)

File: src/main/java/com/bookstore/exception/CustomGlobalExceptionHandler.java (hunk 3 above)
Comment (NikitaLafinskiy) [reply src/main/java/com/bookstore/exception/CustomGlobalExceptionHandler.java:45]: > not fixed

Since we do not have a separate task for it and the change is minimal I decided to put it into this PR, but it does not have any real impact on the way the program runs, so I can also remove this line completely

File: src/main/java/com/bookstore/controller/ShoppingCartController.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/cart")
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+
+    @GetMapping
+    @PreAuthorize("hasRole('USER')")
+    public ShoppingCartDto getCart(@AuthenticationPrincipal String email) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/controller/ShoppingCartController.java]: it's better to provide userId instead of email, because PK id is index and it improves DB performance

File: src/main/java/com/bookstore/controller/ShoppingCartController.java [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    @PreAuthorize("hasRole('USER')")
+    public ShoppingCartDto addCartItem(
+            @AuthenticationPrincipal String email,
+            @RequestBody @Valid AddBookCartRequestDto addBookCartRequestDto) {
+        return shoppingCartService.addCartItem(email, addBookCartRequestDto);
+    }
+
+    @PutMapping("/items/{cartItemId}")
+    @PreAuthorize("hasRole('USER')")
+    public CartItemDto updateCartItem(@RequestBody @Valid UpdateCartRequestDto updateCartRequestDto,
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/controller/ShoppingCartController.java]:
```suggestion
    public ShoppingCartDto updateCartItem(@RequestBody @Valid UpdateCartRequestDto updateCartRequestDto,
```

File: src/main/java/com/bookstore/dto/cart/AddBookCartRequestDto.java [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+
+@Data
+public class AddBookCartRequestDto {
+    @NotNull
+    private Long bookId;
+
+    @NotNull
+    @Min(0)
+    @Max(100)
+    private Integer quantity;
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/dto/cart/AddBookCartRequestDto.java]:
```suggestion
    @Positive
    private int quantity;
```

File: src/main/java/com/bookstore/dto/cart/AddBookCartRequestDto.java [hunk 15]
```diff
@@ -0,0 +1,17 @@
+package com.bookstore.dto.cart;
+
+import jakarta.validation.constraints.Max;
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
+import lombok.Data;
+
+@Data
+public class AddBookCartRequestDto {
+    @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/dto/cart/AddBookCartRequestDto.java:9]:
```suggestion
    @NotNull
    @Positive
```

File: src/main/java/com/bookstore/dto/cart/UpdateCartRequestDto.java [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
+import lombok.Data;
+
+@Data
+public class UpdateCartRequestDto {
+    @NotNull
+    @Max(100)
+    @Min(0)
+    private Integer quantity;
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/dto/cart/UpdateCartRequestDto.java]:
```suggestion
    @Positive
    private int quantity;
```

File: src/main/java/com/bookstore/entity/CartItem.java [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "book_id", nullable = false)
+    private Book book;
+
+    @Column(nullable = false)
+    private Integer quantity;
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/CartItem.java]:
```suggestion
    private int quantity;
```

File: src/main/java/com/bookstore/entity/CartItem.java [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "shopping_cart_id", nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "shopping_cart_id", nullable = false)
```

File: src/main/java/com/bookstore/entity/CartItem.java [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+@Table(name = "cart_items")
+@Getter
+@Setter
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "book_id", nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "book_id", nullable = false)
```

File: src/main/java/com/bookstore/mapper/ShoppingCartMapper.java [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.mapstruct.AfterMapping;
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import org.mapstruct.MappingTarget;
+import org.springframework.beans.factory.annotation.Autowired;
+
+@Mapper(config = MapperConfig.class, uses = { UserMapper.class, CartItemMapper.class })
+public abstract class ShoppingCartMapper {
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/mapper/ShoppingCartMapper.java]:
```suggestion
public interface ShoppingCartMapper {
```

File: src/main/java/com/bookstore/service/cart/ShoppingCartServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+        shoppingCart.addCartItem(cartItem);
+        shoppingCartRepository.save(shoppingCart);
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public CartItemDto updateCartItem(UpdateCartRequestDto updateCartRequestDto,
+                                      Long cartItemId) {
+        CartItem cartItem = cartItemRepository.findById(cartItemId)
+                .orElseThrow(() -> new EntityNotFoundException("Cart item not found"));
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/service/cart/ShoppingCartServiceImpl.java]: provide cartItemId to message

File: src/main/java/com/bookstore/service/user/UserServiceImpl.java [hunk 22]
```diff
@@ -31,6 +36,9 @@ public UserRegistrationResponseDto register(UserRegistrationRequestDto
         User user = userMapper.toEntity(userRegistrationRequestDto);
         user.setPassword(passwordEncoder.encode(userRegistrationRequestDto.getPassword()));
         user.setRoles(Set.of(roleRepository.findByRoleName(DEFAULT_ROLE)));
-        return userMapper.toDto(userRepository.save(user));
+        userRepository.save(user);
+        ShoppingCart shoppingCart = new ShoppingCart(user);
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/service/user/UserServiceImpl.java]: create method in ShopCartService to` createShoppingCart `

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/com/bookstore/mapper/ShoppingCartMapper.java (hunk 20 above)
Comment (NikitaLafinskiy) [reply src/main/java/com/bookstore/mapper/ShoppingCartMapper.java]: I can not understand how to make this into an interface but keep the ShoppingCartMapper and the CartItemMapper logic separated, initially I put it into the same class because I did not want to make the ShoppingCartMapper an abstract class, however it does work as an abstract class as well.
I can not really find any info on how to do that but keep it as an interface, the only way that I've found on the internet is to do the following, but it currently breaks the application, and I am not sure why:

public interface ShoppingCartMapper {
    
    CartItemMapper cartItemMapper(); // apparantly this is supposed to return an instance of CartItemMapper but it can not because it is an interface and not a class

    @Mapping(target = "cartItems", ignore = true)
    @Mapping(target = "userId", source = "user.id")
    ShoppingCartDto toDto(ShoppingCart shoppingCart);

    @AfterMapping
    default void setCartItemDtos(ShoppingCart shoppingCart,
                                 @MappingTarget ShoppingCartDto shoppingCartDto) {
        Set<CartItemDto> cartItemDtos = shoppingCart.getCartItems()
                .stream()
                .map(cartItemMapper()::toDto)
                .collect(Collectors.toSet());
        shoppingCartDto.setCartItems(cartItemDtos);
    }

Comment (ahoienko) [reply src/main/java/com/bookstore/mapper/ShoppingCartMapper.java]: When you used 'uses' in Mapper annotation framework will create method similar to 'setCartItemDtos' so there is no reason to autowire anything. So you can create an interface

**Review by ahoienko — CHANGES_REQUESTED** [review]
Nice, left one commnet

**Review by Elena-Bruyako — APPROVED** [review]
